### PR TITLE
[FLINK-16770][checkpoint] Avoid dispose checkpoint during async finalization of pendingcheckpoint

### PR DESCRIPTION
**Pull request by Myasuka**
## What is the purpose of the change

This is a quick hot-fix of FLINK-16770, which would not remove checkpoint by mistake but would let checkpoint could complete after job switched to cancelling status. However, this should not introduce harmful influence.

## Brief change log

Refactor the logic of `PendingCheckpoint#dispose` and `PendingCheckpoint#finalizeCheckpoint` so that they are exclusive from each other. 

## Verifying this change

This change added tests and can be verified as follows:

  - Refactor `PendingCheckpoint#testAbortingDuringAsyncFinalization` to verify not discard checkpoint when already in async phase of finalize checkpoint.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers:no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by ifndef-SleePy** [comment]
Hi @Myasuka , there are several unit testing cases failed. I think we should fix them first.

**Comment by Myasuka** [comment]
@ifndef-SleePy Already fix the broken tests by allowing pending checkpoint could dispose after finalization failed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java [hunk 1]
```diff
@@ -602,6 +624,17 @@ private void reportFailedCheckpoint(Exception cause) {
 		}
 	}
 
+	private boolean grabStatusPriority() {
```
Comment (ifndef-SleePy) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java:627]: Could you think of a more straightforward name? It's a kind of releasing mutex here, "priority" is not accurate.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java [hunk 2]
```diff
@@ -129,6 +127,12 @@
 
 	private CheckpointException failureCause;
 
+	/**
+	 * The CAS flag for avoiding conflict of dispose and finalize operations.
+	 * Who grab this flag would avoid the other to enter its operation.
+	 */
+	private final AtomicBoolean statusFlag;
```
Comment (ifndef-SleePy) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java:134]: The name of `statusFlag` is a bit general.

**Comment by Myasuka** [comment]
Close this PR as we finally revert all related commits.

**Pull request closed without merging**